doc/SAI-Proposal-SAI-PTF.md

```diff
@@ -0,0 +1,627 @@
+| **Templates** <br /> - *sai_rpc_server_functions.tt* <br /> - *sai_rpc_server_helper_functions.tt* <br /><br />**Auto-generated templates** <br /> - *sai_rpc_server.cpp.tt* (by Thrift from *sai.thift*) <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai_rpc_server.cpp** <br /> - thrift server <br /> - need *sai.thrift* and *Thrift* itself for generation |
+| **Templates** <br /> - *sai_adapter.py.tt* <br /> - *sai_adapter_utils.tt* <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai_adapter.py** <br /> - thrift client and main SAI PTF test library <br /> - generated from scratch |
+
+
+All of this can be generated with gensairpc.pl script.
+
+![autogeneration parsing](figures/autogeneration_parsing.png "Figure 2: Autogeneration - parsing")
+__Figure 2: Autogeneration - parsing__
+
+![autogeneration gensairpc](figures/autogeneration_gensairpc.png "Figure 3: Autogeneration - gensairpc")
+__Figure 3: Autogeneration - parsing__
+
+Autogeneration:
+
+```
+cd <root>/meta/
+./gensairpc.pl --clean
+```
+
+When upgrading SAI:
+
+```
+cd <root>/meta
+make
+./gensairpc.pl --clean --no-meta-build
+make clean - C <root>/meta
+```
+
+Use --no-meta-build because SAI meta are already built with make.
+
+## Test execution
+
+Test files written in Python call functions from sai_adapter.py. When running such test auto-generated Python functions are talking with C-based RPC server to run C functions. These functions are used to create/remove SAI object and get/set their attributes. There are also functions allowing to read object statistics and bulk functions. Then it is running internal SAI implementation adopting common SAI interface to test the implementation.
+
+Tests use SAI to configure underlay switch. Using PTF we send and verify packets that are testing switch configuration.
+
+## Tests structure
+
+The tests themselves are based on python unittest framework (https://docs.python.org/3/library/unittest.html). They consist of three main functions:
+
+- setUp() - creating necessary objects
+- runTest() - running proper test
+- tearDown() - removing all created objects
+
+Tests have to inherit from one of two base classes - SaiHelper or SaiHelperBase. SaiHelperBase is the base class that runs initial configuration for all tests. 
+
+SaiHelperBase initializes switch, gets several switch attributes and stores them into class attributes that are later to be used by tests. The attributes that are stored are:
+- default_vlan_id
+- default_vrf
+- default_1q_bridge
+- acl_stage_ingress
+- acl_stage_egress
+- number_of_active_ports
+- port_list stored into variables self.portX
```

Comment:
I would say port numbers instead of port_list - this looks like a variable name. Also, 'self' is unnecessary if we don't use this keyword nowhere else.